hmkit-android/src/main/java/com/highmobility/hmkit/GattServer.java

```diff
@@ -322,6 +356,75 @@ public void onCharacteristicReadRequest(final BluetoothDevice device,
                 offset,
                 offsetBytes);
 
+
+        /**
+         * Offset problem detection logic
+         *
+         * If detection is enabled
+         */
+        if (configuration.bleEnableOffsetProblemDetection()) {
+
+            boolean detectedOffsetProblem = false;
+
+            /**
+             * If value length is reached and read request comes still in
+             * more than 5 times then set flag that problem is detected
+             */
+            offsetFixlastOffset = offset;
+            if(offsetFixlastOffset == value.length){
+                offsetFixlastOffsetCount++;
+                if(offsetFixlastOffsetCount > 5){
+                    detectedOffsetProblem = true;
+                }
+            }else{
+                offsetFixlastOffsetCount = 0;
+            }
+
+            /**
+             * If problem is detected then check if preferences needs to be changed
+             */
+            if(detectedOffsetProblem == true){
+                boolean storeToShared = false;
+                Integer sharedEnableValue = 0;
+                /**
+                 * If sharedOffsetFixEnabled is 0
+                 * it means values are not set to shared preferences
+                 * and enable offset fix logic
+                 */
+                if(sharedOffsetFixEnabled == 0){
+                        storeToShared = true;
+                        sharedEnableValue = 1;
+                }else{
+                    /**
+                     * If problem is detected and values are set to shared preferences
+                     * then change them only when os release version is changed
+                     */
+                    if(sharedOffsetFixOsRelease != Build.VERSION.RELEASE){
+                        storeToShared = true;
+                        if(sharedEnableValue == 1){
+                            sharedEnableValue = 2;
+                        }else{
+                            sharedEnableValue = 1;
+                        }
+                    }
+                }
+
+                /**
+                 * Store offset problem detection values to shared preferences
+                 */
+                if(storeToShared == true){
+                    SharedPreferences sharedPref = context.getSharedPreferences(offsetFixSharedPreferences, MODE_PRIVATE);
```

Comment:
Should we move SharedPreferences callsto Storage.java class? That class already has SharedPreferences management.